samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -443,6 +443,18 @@ public void run() {
             long startTime = clock.nanoTime();
             task.window(coordinator);
             containerMetrics.windowNs().update(clock.nanoTime() - startTime);
+
+            // We have scenario when the task window method takes longer time than window trigger time, there is a
+            // potential for process method starvation. Ideally, we need to merge the thread pool for commit, window with
+            // process to guarantee fairness. In the interim, we want to warn users about potential starvation.
+            double lowerBoundForWindowTriggerTimeInMs =
+                containerMetrics.windowNs().getSnapshot().getAverage() / 1e6;
+            if (windowMs <= lowerBoundForWindowTriggerTimeInMs) {
+              log.warn(
```

Comment:
I am still going to leave it as potential starvation scenario since the upcoming window method could end up not starving the process method. I will update the warn statement to use task.window.ms